operators/nftopology/controllers/nftopology_controller.go

```diff
@@ -0,0 +1,182 @@
+	upstream.Revision = nfClassObj.Spec.PackageRef.Revision
+
+	retPVS.Spec.Upstream = upstream
+
+	target := &pvsv1alpha2.Target{}
+	objectSelector := &pvsv1alpha2.ObjectSelector{}
+
+	labelMap := map[string]string{}
+	for k, v := range nfInst.ClusterSelector.MatchLabels {
+		labelMap[k] = v
+	}
+
+	objectSelector.LabelSelector = metav1.LabelSelector{MatchLabels: labelMap}
+	objectSelector.APIVersion = WORKLOADAPIVERSION
+	objectSelector.Kind = WORKLOADKIND
+	target.ObjectSelector = objectSelector
+
+	retPVS.Spec.Targets = append(retPVS.Spec.Targets, *target)
+
+	// TODO(s3wong): adding Template in PVS?
+	return retPVS
+}
+
+//+kubebuilder:rbac:groups=req.nephio.org,resources=nftopologies,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=req.nephio.org,resources=nftopologies/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=req.nephio.org,resources=nftopologies/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *NFTopologyReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	r.l = log.FromContext(ctx).WithValues("req", req)
+	r.l.Info("Reconcile NFTopology")
+
+	nfTopo := &reqv1alpha1.NFTopology{}
+	if err := r.Client.Get(ctx, req.NamespacedName, nfTopo); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	nfDeployed := &deployv1alpha1.NFDeployed{}
+	if err := r.Client.Get(ctx, client.ObjectKey{Namespace: nfTopo.ObjectMeta.Namespace, Name: nfTopo.ObjectMeta.Name}, nfDeployed); err != nil {
+		r.l.Info(fmt.Sprintf("NF Deployed object for %s not created, continue...\n", nfTopo.ObjectMeta.Name))
+	}
+
+	neighborMap := BuildNeighbormap(BuildAttachmentMap(nfTopo))
+	deployedInstanceMap := BuildDeployedInstanceMap(nfDeployed)
+	continueReconciling := false
+	for _, nfInst := range nfTopo.Spec.NFInstances {
+
+		className := nfInst.NFTemplate.ClassName
+		nfClass := &reqv1alpha1.NFClass{}
+		if err := r.Client.Get(ctx, client.ObjectKey{Name: className}, nfClass); err != nil {
+			r.l.Error(err, fmt.Sprintf("NFClass object not found: %s: %s\n", className, err.Error()))
+			return ctrl.Result{}, err
+		}
+		// TODO(s3wong): turn the following into a configurable policy
+		if nfClass.Spec.Vendor == FREE5GC && nfInst.NFTemplate.NFType == reqv1alpha1.NFTypeSMF {
+			r.l.Info(fmt.Sprintf("NF instance free5gc %s is a SMF, checking all connected UPF deployed\n", nfInst.Name))
+			if !AllUPFsDeployed(neighborMap, &nfInst, nfDeployed, deployedInstanceMap) {
+				continueReconciling = true
+				// skip this SMF
+				continue
+			}
+		}
+
+		pvs := BuildPVS(&nfInst, nfClass)
+		r.l.Info(fmt.Sprintf("PVS for NF inst %s is %+v\n\n\n", nfInst.Name, pvs))
+		if err := r.Client.Create(ctx, pvs); err != nil {
+			r.l.Error(err, fmt.Sprintf("Failed to create PVS %s: %s\n", nfInst.Name, err.Error()))
+		} else {
```

Comment:
I'm confused by this else - isn't this saying: if I have an error during create, log it; otherwise update the thing I just created.